vpr/src/backend/physical_netlist_writer.cpp

```diff
@@ -0,0 +1,732 @@
+            VTR_ASSERT(branch.segment == pb_route_branch.segment);
+
+            branch.branches = pb_route_branch.branches;
+
+            phys_nets_[net_id] = route_branch;
+        }
+    }
+}
+
+void InterchangeWriterVisitor::visit_clb_impl(ClusterBlockId blk_id, const t_pb* clb) {
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& device_ctx = g_vpr_ctx.device();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    VTR_ASSERT(clb->pb_graph_node != nullptr);
+    VTR_ASSERT(clb->pb_graph_node->pb_type);
+
+    curr_pb_route_ = clb->pb_route;
+
+    int x = place_ctx.block_locs[blk_id].loc.x;
+    int y = place_ctx.block_locs[blk_id].loc.y;
+    int sub_tile = place_ctx.block_locs[blk_id].loc.sub_tile;
+    auto& grid_loc = device_ctx.grid[x][y];
+
+    physical_tile_ = grid_loc.type;
+    logical_block_ = cluster_ctx.clb_nlist.block_type(blk_id);
+
+    auto tileTypeList = dr_.getTileTypeList();
+    auto siteTypeList = dr_.getSiteTypeList();
+
+    bool is_const = std::string(physical_tile_->name) == "constant_block";
+    if (is_const)
+        return;
+
+    VTR_ASSERT(grid_loc.meta != nullptr && grid_loc.meta->has(ic_gridloc));
+
+    auto* value = grid_loc.meta->get(ic_gridloc);
+    VTR_ASSERT(value != nullptr);
+
+    auto sites_info = vtr::split(value->front().as_string().get(strings_), ",");
+    VTR_ASSERT(sub_tile < (int)sites_info.size());
+
+    auto site_info = vtr::split(sites_info[sub_tile], ":");
+    auto site_name = current_site_ = std::stoi(site_info[0]);
+
+    auto site_type_idx = std::stoi(site_info[1]);
+    auto tile_type_idx = std::stoi(site_info[2]);
+
+    auto tile_type = tileTypeList[tile_type_idx];
+    current_site_type_ = tile_type.getSiteTypes()[site_type_idx].getPrimaryType();
+    auto site_type = siteTypeList[current_site_type_];
+
+    site_instances_.emplace_back(site_name, site_type.getName());
+
+    auto& pb_type = logical_block_->pb_type;
+
+    std::vector<t_pb_graph_pin*> pb_pins;
+    int out_port_index = 0;
+    for (int iport = 0; iport < pb_type->num_ports; iport++) {
+        if (pb_type->ports[iport].type != OUT_PORT)
+            continue;
+
+        for (int ipin = 0; ipin < pb_type->ports[iport].num_pins; ipin++)
+            pb_pins.push_back(&clb->pb_graph_node->output_pins[out_port_index++][ipin]);
+    }
+
+    iterate_nodes(pb_pins, true);
+}
+
+void InterchangeWriterVisitor::visit_all_impl(const t_pb_routes& pb_routes, const t_pb* pb) {
+    VTR_ASSERT(pb != nullptr);
+    VTR_ASSERT(pb->pb_graph_node != nullptr);
+}
+
+void InterchangeWriterVisitor::visit_route_through_impl(const t_pb* atom) {
+    VTR_ASSERT(atom != nullptr);
+}
+
+std::unordered_map<size_t, size_t> InterchangeWriterVisitor::get_lut_pin_rotation(const t_pb* atom, bool is_lut) {
+    auto& nlist = g_vpr_ctx.atom().nlist;
+
+    auto pb_graph_node = atom->pb_graph_node;
+    auto* pb_type = pb_graph_node->pb_type;
+
+    int iport = 0; // Lut input port is at index 0
```

Comment:
If are making such assumptions (and if there's no other way to identify LUT input port), there should be a #define/const/constexpr for that, eg. `LUT_INPUT_PORT`. Whatever code assigns the input port to`pb_type` for a LUT, should make an assert that it is indeed equal to `LUT_INPUT_PORT`. This will make future developers (or the current ones, if they forget about it) aware of this assumption whenever they wan't to make some changes to LUT-related code.

Reply:
Right, added a `const` to use both here and in the arch reader